Bug fix for Brexit checker action groupings

hannako opened the pull request:
## What
Fix two bugs on the Brexit checker results page.

### Bug one
A user's selected criteria results in an action that has two grouping criteria. We always show both groups, regardless of the selected criteria. eg A user selects `living-uk`, `visiting-ie` and `visiting-bring-pet`.  Because the action for taking a pet abroad has both `visiting-eu` and `visiting-ie` as grouping criteria, the user is shown the `visiting-ie` group (makes sense) and also the `visiting-eu` group (whoops!). 

- [bug on production](https://www.gov.uk/transition-check/results?c%5B%5D=does-not-own-operate-business-organisation&c%5B%5D=visiting-bring-pet&c%5B%5D=visiting-ie&c%5B%5D=living-uk)

#### New logic:
- If an action has more than one grouping criteria, and a user selects
all of the ANY_OF criteria, we should show duplicate actions under both groups.
- If an action has more than one grouping criteria, and a user selects
some of the ANY_OF criteria, we should only show the action under the grouping that
has a key in common with the user's selected criteria.

- [fixed on review app](https://finder-front-remove-dup-ynsole.herokuapp.com/transition-check/results?c%5B%5D=does-not-own-operate-business-organisation&c%5B%5D=visiting-bring-pet&c%5B%5D=visiting-ie&c%5B%5D=living-uk)

### Bug two
When an action has two grouping criteria, and both criteria have been selected by the user, we display both criteria under each group. Eg a user selects `visiting-bring-pet` `visiting-ie` `visiting-eu` `living-uk`. This gives them the following grouped actions:

**Visiting the EU**
Because you said:

You live in the UK
You plan to travel to Ireland
You plan to travel to the EU
You are travelling with a pet

**Visiting Ireland**
Because you said:

You live in the UK
You plan to travel to Ireland
You plan to travel to the EU
You are travelling with a pet

-[bug on production](https://www.gov.uk/transition-check/results?c%5B%5D=does-not-own-operate-business-organisation&c%5B%5D=visiting-bring-pet&c%5B%5D=visiting-ie&c%5B%5D=visiting-eu&c%5B%5D=living-uk)

#### New logic:
Previously we were creating the list of criteria by combining all of the criteria for all of the actions displayed in a group. Now if an action has multiple grouping criteria, we check if the action's criteria shares a key with the current group. If it does we display the criteria, if it does not we don't.

-[fixed on review app](https://finder-front-remove-dup-ynsole.herokuapp.com/transition-check/results?c%5B%5D=does-not-own-operate-business-organisation&c%5B%5D=visiting-bring-pet&c%5B%5D=visiting-ie&c%5B%5D=visiting-eu&c%5B%5D=living-uk)

:warning: This application is Continuously Deployed: :warning:

- Merged changes are automatically deployed to staging and production.

- Make sure you follow [the guidance for deployments](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) **before** you merge.

- Check your branch is being deployed in the [Release app](https://release.publishing.service.gov.uk/applications/finder-frontend), after merging.

https://trello.com/c/MufPWtKB/376-investigate-how-we-can-remove-duplicate-actions-from-the-results-page

hannako commented on app/lib/brexit_checker/group.rb:
```diff
@@ -15,7 +15,7 @@ class BrexitChecker::Group
                                 studying-uk
                                 common-travel-area]
 
-  attr_reader :key, :heading
```
The priority of each group is set in the [spreadsheet](https://docs.google.com/spreadsheets/d/1wIeBTitJVfkWa7oKrGmusIo2r4TsvXVdlne_xG6YjYs/edit#gid=443740486) but I can't see anywhere in the app that the priority is pulled in and used to set the order of the groups. It looks like the order of the elements in the yaml file is all that was previously determining the group order.

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+          multiple_grouping_criteria?(action) ? selected_group(action, selected_criteria) : action.grouping_criteria
+        group_keys.each do |key|
+          group = all_groups.select { |g| g.key == key }.first
+          next if grouped_actions.select { |actions| actions[:group] == group }.any?
```
you can collapse this to be `next if grouped_actions.any? { |actions| actions[:group] == group }` I think, and it should have the same advantages as `detect` above :-)

hannako replied:
thank you!

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+            actions: actions_for_group(audience_actions, group),
+            criteria: criteria_for_group(audience_actions, group, selected_criteria),
           }
+          grouped_actions.sort_by! { |actions| [(actions[:group].priority * -1), actions[:group].key] }
```
Feels like the sorting by priority is significant enough to have a commit of its own? This is quite complicated :-)

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+      filtered_criteria =
+        actions.map do |action|
+          if multiple_grouping_criteria?(action)
+            rogue_criterion = (action.grouping_criteria - [group.key]).first
```
This feels quite magic to me - could you walk me through it? I don't feel confident enough to know whether the `first` criterion is the one we need, or whether they may be others that we're ignoring, or I've missed something completely.

hannako replied:
Currently the highest number of grouping criteria an action can have is 2, so calling `first` was just a way to access the element from the array. But your comment made me think that if in the future an action is given more than 2, this code would have introduced a bug. So I've reworked in a later commit:

```
def filtered_criteria(action, group_key)
  unfiltered = criteria_for_action(action)
  rogue_criteria_keys = (action.grouping_criteria - [group_key])
  unfiltered.reject { |criterion| rogue_criteria_keys.include? criterion.key }
end
```

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
       end
     end
+
+    def selected_group(action, selected_criteria)
```
Is this named correctly? It seems to be more `grouping_criteria` rather than `group`, or something?

hannako replied:
I've renamed `selected_grouping_criteria`, as it's the grouping criteria which have a key in common with the user's selected criteria, rather than all of that action's `grouping_criteria`.

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+    end
+
+    def criteria_for_group(audience_actions, group, selected_criteria)
+      actions = audience_actions & group.actions
```
you could reuse the `actions_for_group` method you've created a couple of lines up? ;-)

hannako replied:
thanks 👍

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+            rogue_criterion = (action.grouping_criteria - [group.key]).first
+            (action.all_criteria & selected_criteria).reject { |criteria| criteria.key == rogue_criterion }
+          else
+            action.all_criteria & selected_criteria
```
you've used this in a couple of places here. Maybe extract a method, so you could name your intent?

sihugh commented on spec/lib/brexit_checker/results_audiences_spec.rb:
```diff
@@ -2,81 +2,121 @@
-
-    let(:selected_criteria) { [criteria1, criteria2, criteria3, criteria4, criteria5] }
-    let(:actions) { [action1, action2, action3, action4, action5, action6] }
+    living_uk = FactoryBot.build(:brexit_checker_criterion, key: "living-uk")
```
Was there a reason why these are no longer `let(:living_uk)` etc.?

hannako replied:
Every time factorybot builds the criteria or action via the let, it gets a new id. Because these models aren't active record objects, they are just hashes, if the value of the id changed the test failed. 

Because the criteria and actions are loaded from the yaml more than once, and passed around, even though the data is consistent the ID is changing. Which is maybe bad? Makes it hard to test at the very least. So in a later commit I refactored so that the criteria and actions are passed in at the initialise, so they remain unchanged - I reinstate the `let` blocks at that point.

sihugh commented on app/controllers/brexit_checker_controller.rb:
```diff
@@ -50,6 +50,10 @@ def confirm_email_signup
 private
 
+  def audience_results
+    BrexitChecker::ResultsAudiences.new(@audience_actions, @criteria)
```
Would this benefit from memoization?

sihugh commented on app/lib/brexit_checker/action.rb:
```diff
@@ -62,6 +62,10 @@ def eql?(other)
   end
 
+  def multiple_grouping_criteria?
+    grouping_criteria.count > 1
```
👍

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -1,65 +1,68 @@
-  class << self
-    def populate_business_groups(audience_actions, selected_criteria)
-      return {} if audience_actions.blank? || selected_criteria.blank?
+  attr_reader :citizen_actions, :business_actions, :criteria
```
This is a nice thing to do 👍

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -1,65 +1,68 @@
-        end
-      filtered_criteria.flatten.uniq
-    end
+  def criteria_keys
```
👍

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -1,65 +1,68 @@
-    def multiple_grouping_criteria?(action)
-      action.grouping_criteria.count > 1
-    end
+  def criteria_for_group(group)
```
wonder if it's worth naming this as being specific to citizens?

hannako replied:
I've refactored so that the actions are passed in.  If we add groupings for business actions we can reuse the method, in theory!

sihugh commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -1,65 +1,68 @@
-        group_keys.each do |key|
-          group = all_groups.select { |g| g.key == key }.first
-          next if grouped_actions.select { |actions| actions[:group] == group }.any?
+  def populate_citizen_groups
```
This method is still pretty complicated. Any chance we could break it down a little?

sihugh reviewed (changes requested):
I'm pretty sure you've got the right outcome, but I'm afraid I found the 3rd commit too big for my brain to cope with right now. Could we have a go at splitting it out a bit? Happy to pair on it.

hannako commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+        next if action.grouping_criteria.empty?
+
+        group_keys =
+          multiple_grouping_criteria?(action) ? selected_group(action, selected_criteria) : action.grouping_criteria
```
The grouping criteria key on the groups.yaml file stores an array. So even if it only has one grouping_criteria, it's still an array of one element. [eg](https://github.com/alphagov/finder-frontend/blob/314aa1112d0333cf44f947e52b47b98dc712349d/app/lib/brexit_checker/actions.yaml#L21)

hannako commented on app/lib/brexit_checker/results_audiences.rb:
```diff
@@ -12,17 +12,54 @@ def populate_business_groups(audience_actions, selected_criteria)
+        group_keys =
+          multiple_grouping_criteria?(action) ? selected_group(action, selected_criteria) : action.grouping_criteria
+        group_keys.each do |key|
+          group = all_groups.select { |g| g.key == key }.first
```
brilliant thank you!

hannako commented:
Thank you for the review! I've split the third commits into smaller pieces, and I've extracted the criteria filtering to its own class. Hopefully that makes it a bit clearer.

sihugh reviewed (approved):
Looks grand 👍 🎉 🚀 🦸 

Do we need to get Marcus Arborius to re-review?
